### ISSUE TITLE
Feature Request: Make CIntervals awaitable

### ISSUE THREAD

**Issue by LittleToonCat** [panda3d/panda3d#909]
Recently, I've found this [old blog post about coroutines (async/await) support](https://www.panda3d.org/blog/november-2017-development-update/) and I have noticed that Intervals are supported (`await interval.start()`).  When I first read about it, I thought *any* interval can be awaited, but when I tried awaiting a Sequence, Parallel, or even an Actor's pos/hpr intervals, it wouldn't work.

After doing some source digging, I've found out that only Python based intervals can be awaited (ActorInterval, SoundInterval, etc.) and not intervals that are based on CInterval, that's including MetaIntervals (Sequence, Parallel), and LerpIntervals (actor.posHprInterval). Such intervals people might use a lot in a typical Panda game project.

It will be nice to have these intervals awaitable to have more finer control over them.

**Comment by rdb** [comment]
You're right, this is an important omission, there is no good reason why CInterval should not be awaitable as well.

**Comment by rdb** [comment]
I need an opinion on what would be most intuitive: if some code is awaiting an interval that is

**Comment by rdb** [comment]
Whoops, I pressed a wrong button.

I need an opinion on what would be the most intuitive: if some code is awaiting an interval and some other code pauses or stops it, what should happen to the awaiting code?  Should it continue waiting until at some point some other code continues it (which may never happen)?  Or should it throw a CancelledException into the awaiting code?

I think that if other code calls `finish()` on the interval, the awaiting coroutine should resume as if the interval had run its course, but I am personally leaning towards having any other interruption (including pause) throw some kind of "interval interrupted" exception that the awaiting code can handle as it pleases.

**Comment by Moguri** [comment]
I also lean toward throwing an exception. If we really wanted the option of not having an an exception, we could add an argument to `start()` to not throw exceptions on interrupts (default to throw). I ran into this the other day when trying to await a Sequence() can got rather confused to get `None` back.

**Comment by rdb** [comment]
Yeah, I think if someone intends to cancel the interval, then a CancelledException should propagate into the coroutine.  If your coroutine awaits some interval, the code following the `await` it is likely to depend on the effects of it.  (This behaviour would be analogous to stopping a Sequence; when you do so, the intervals following the current one will never play either.)

But when you call `pause()` eg. as part of entering a "Pause" game state that pauses all active intervals, it seems more likely that the user would want the coroutine to continue awaiting until the interval is unpaused.

Unfortunately, the way intervals work, there's no `cancel()` or `stop()` operation—if you want to pause an interval, you currently just call `pause()` (which removes the reference to it from the interval manager) and then let it go out of scope.  So we don't currently let the user describe this intention clearly enough to make this distinction.

I would propose the following:
* We add a `cancel()` or `stop()` method to all intervals.  This would internally call `pause()` but also throw a `CancelledException` into any awaiting coroutines.  We recommend that users use this to permanently abort their intervals.
* `pause()` will do nothing to the coroutine; it continues awaiting the interval, since it may be resumed later.
* If a paused interval goes out of scope while a coroutine is still awaiting it, we throw a special kind of `CancelledException` telling the user that they should have used `cancel()`/`stop()`.
* `finish()` does not throw any exceptions, because this operation just fast-forwards the interval to the end, so the coroutine can just continue to run.

In any case, I don't think we need an explicit option to suppress, since the user can use a `try..except` block or use a wrapper that suppresses the exception (like `asyncio.shield()`).

**Comment by rdb** [comment]
I implemented this for 1.10.9, though somewhat inefficiently (it doesn't suspend the task, but polls the task every frame until the interval is done), and without cancellation support.  I'm leaving this open to revisit for 1.11.
